xapian-core/include/xapian/diversify.h

```diff
@@ -0,0 +1,134 @@
+#endif
+
+#include <xapian/attributes.h>
+#include <xapian/cluster.h>
+#include <xapian/mset.h>
+#include <xapian/types.h>
+#include <xapian/visibility.h>
+
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+namespace Xapian {
+
+/** Class for diversifying an MSet using GLS-MPT as given
+ *  in the paper: Scalable and Efficient Web Search Result
+ *  Diversification Naini et al. 2016
+ */
+class XAPIAN_VISIBILITY_DEFAULT Diversify {
+    /// Top-k documents of given mset are diversified
+    unsigned int k;
+
+    /// MPT parameters
+    double lambda, b, sigma_sqr;
+
+    /// Store each document from given mset as a point
+    std::unordered_map<Xapian::docid, Xapian::Point> points;
+
+    /// Store the relevance score of each document
+    std::unordered_map<Xapian::docid, double> scores;
+
+    /// Store pairwise cosine similarities of documents of given mset
+    std::map<std::pair<Xapian::docid, Xapian::docid>, double> pairwise_sim;
+
+    /// Store docids of top k diversified documents
+    std::vector<Xapian::docid> main_dmset;
+
+    /** Initial diversified document set
+     *
+     *  Return top-k documents of mset as vector of Points, which
+     *  represent the initial diversified document set.
+     *
+     *  @param source	MSet object containing the documents of which
+     *			top-k are to be diversified
+     */
+    void initialise_points(const Xapian::MSet& source);
+
+    /** Return a key for a pair of documents
+     *
+     *  Returns a key as a pair of given documents ids
+     *
+     *  @param source	MSet object containing the documents of which
+     *			top-k are to be diversified
+     */
+    std::pair<Xapian::docid, Xapian::docid>
+    get_key(const Xapian::docid& docid_a, const Xapian::docid& docid_b);
+
+    /** Compute pairwise similarities
+     *
+     *  Used for pre-computing pairwise cosine similarities of documents
+     *  of given mset, which is used to speed up evaluate_dmset
+     */
+    void compute_similarities();
+
+    /** Return difference of 'points' and current dmset
+     *
+     *  Return the difference of 'points' and the current diversified
+     *  document match set
+     *
+     *  @param dmset	Document set representing a diversified document set
+     */
+    std::vector<Xapian::docid>
+    compute_diff_dmset(const std::vector<Xapian::docid>& dmset);
+
+    /** Evaluate a diversified mset
+     *
+     *  Evaluate a diversified mset using MPT algorithm
+     *
+     *  @param dmset	Set of points representing candidate diversifed
+     *			set of documents
+     */
+    double evaluate_dmset(const std::vector<Xapian::docid>& dmset);
+
+  public:
+    /** Constructor specifying the number of diversified search results
+     *
+     *  @param  k_	Number of required diversified results
+     */
+    explicit Diversify(unsigned int k_,
+		       double lambda_ = 0.5,
+		       double b_ = 5.0,
+		       double sigma_sqr_ = 1e-3);
```

Comment:
It'd be good to document the other parameters too.

Reply:
Thanks for pointing this out. Modifications pushed in a fixup commit.